pkg/tenant/topics.go

```diff
@@ -237,24 +254,100 @@ func (s *Server) TopicUpdate(c *gin.Context) {
 		return
 	}
 
+	// Sanity check that the ID in the URL matches the ID in the request body.
+	if topic.ID != topicID.String() {
+		log.Warn().Msg("topic id in request body does not match topic id in URL")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("id in request body does not match id in URL"))
+		return
+	}
+
 	// Verify the topic name exists and return a 400 response if it doesn't.
 	if topic.Name == "" {
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic name is required"))
 		return
 	}
 
-	// Get the specified topic from the database and return a 404 response if
-	// it cannot be retrieved.
+	// Topic state is optional but can only be set to readonly.
+	if topic.State != "" && topic.State != pb.TopicTombstone_READONLY.String() {
```

Comment:
Potentially this is too confusing to use and we should just have an archive or readonly flag in the request?

Reply:
Hmm, good point. The action of archiving makes the topic readonly -- and there is currently no mechanism to "unarchive" a topic (although we could potentially add this in the future). However, the state of the topic can change in Ensign and that might need to be reported back to the user. One thing we have to be careful of is to ensure that Tenant and Ensign are synchronized so that user actions in Tenant are reflected globally across the system. 

One way we could decouple this is by making archiving a distinct action rather than an update; e.g. having an archive endpoint; e.g. `POST /v1/topics/:topicID/archive`. This makes the action a little bit stronger and reduces the logic required for determining what the user wants to do. 

If we don't want another endpoint - then the logic for `PUT` is  `t.State != READONLY && topic.State != t.State` where `t` is the topic retrieved from the database. The semantic for `PUT` is that the user has to supply the entire document even if there are fields that they don't want to change (unlike `PATCH` where users specify only the fields they want to change).